src/raft.rs

```diff
@@ -1851,34 +1839,35 @@ impl<T: Storage> Raft<T> {
     /// Indicates whether state machine can be promoted to leader,
     /// which is true when its own id is in progress list.
     pub fn promotable(&self) -> bool {
-        self.prs().voters().contains_key(&self.id)
+        self.prs().voter_ids().contains(&self.id)
     }
 
-    fn add_voter_or_learner(&mut self, id: u64, is_learner: bool) {
-        if self.prs().voters().contains_key(&id) {
-            if is_learner {
-                info!(
-                    "{} ignored add learner: do not support changing {} from voter to learner",
-                    self.tag, id
-                );
-            }
-            // Ignore redundant add voter.
-            return;
-        } else if self.prs().learners().contains_key(&id) {
-            if is_learner {
-                // Ignore redundant add learner.
+    fn add_voter_or_learner(&mut self, id: u64, learner: bool) {
+        debug!(
+            "Adding node (learner: {}) with ID {} to peers.",
+            learner, id
+        );
+        let progress = Progress::new(self.raft_log.last_index(), self.max_inflight, learner);
+        // Ignore redundant inserts.
+        if let Some(progress) = self.prs().get(id) {
+            if progress.is_learner == learner {
                 return;
             }
-            if let Err(e) = self.mut_prs().promote_learner(id) {
-                panic!("{}", e)
-            }
-            if id == self.id {
-                self.is_learner = false;
-            }
+        };
+
+        let result = if learner {
+            self.mut_prs().insert_learner(id, progress)
+        } else if self.prs().learner_ids().contains(&id) {
+            self.mut_prs().promote_learner(id)
         } else {
-            // New progress.
-            let last_index = self.raft_log.last_index();
-            self.set_progress(id, 0, last_index + 1, is_learner);
+            self.mut_prs().insert_voter(id, progress)
+        };
+
+        if let Err(e) = result {
+            panic!("{}", e)
```

Comment:
This seems a broken change. In the past, when trying to downgrade a voter to learner, it will be simply ignored.

Reply:
Panic may prevent rolling update if the downgrade is decided to be supported in the future.

Reply:
Hm, I'd like to eventually make these return a `Result` so this can be handled outside the library.